File: Binary_Search_Tree/701.Insert_into_a_Binary_Search_Tree/Python/701.Insert_into_a_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def insertIntoBST(self, root, val):
        """
        :type root: Optional[TreeNode]
        :type val: int
        :rtype: Optional[TreeNode]
        """
        #==================================#
        # Recursion-based traversal method #
        #==================================#
        if (not root): #Issue/Boundary-case handling
            return TreeNode(val)

        ################################
        #Recursion-based loop traversal
        ##### Check if the current inserted-value matched conditions or not #####
        if ((root.val) >= val):
            root.left = self.insertIntoBST(root.left, val) #Recursion function call (i.e. root left-half node)
        else:
            root.right = self.insertIntoBST(root.right, val) #Recursion function call (i.e. root right-half node)

        return root

File: Binary_Search_Tree/701.Insert_into_a_Binary_Search_Tree/C++/701.Insert_into_a_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode * insertIntoBST(TreeNode * root, int val) 
    {
        //========================================//
        // Input type:                            //
        //   - TreeNode * (root tree-node)        //
        //   - int (inserted-value)               //
        // Return type:                           //
        //   - TreeNode * (result root tree-node) //
        //========================================//
        if (! root) { return new TreeNode(val); } //Issue, Boundary-case handling

        /*Recursion-based loop traversal*/
        ///// Check if the current inserted-value matched conditions or not /////
        if ((root->val) >= val)
        {
            (root->left) = insertIntoBST((root->left), val); //Recursion function call (i.e. root left-half node)
        }
        else
        {
            (root->right) = insertIntoBST((root->right), val); //Recursion function call (i.e. root right-half node)
        }

        return root;
    }
};
